File: vector/RotatedSortedArray.py
from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l = 0
        r = len(nums)
        while l < r:
            mid = (l + r) // 2
            if nums[mid] == target:
                return mid
            if nums[l] < nums[mid]:
                if nums[l] <= target and nums[mid] > target:
                    r = mid
                else:
                    l = mid + 1
            else:
                if nums[r-1] >= target and nums[mid] < target:
                    l = mid + 1
                else:
                    r = mid
        return -1


nums = [4,5,6,7,0,1,2]
nums = [0,1,2,4,5,6,7]
nums = [1,2,4,5,6,7,0]
s = Solution()
res = s.search(nums, 7)
print(res)

File: vector/RotatedSortedArray.cpp
#include <iostream>
#include <vector>

using std::vector;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size();
        int mid;
        while (left < right) {
            mid = (left + right) / 2;
            if (nums[mid] == target) return mid;
            if (nums[left] <= nums[mid]) {
                if (nums[mid] > target && nums[left] <= target)
                    right = mid;
                else
                    left = mid + 1;
            }
            else {
                if (nums[mid] < target && nums[right-1] >= target)
                    left = mid + 1;
                else
                    right = mid;
            }
        }
        return -1;
    }
};

int main()
{
    Solution s;
    std::vector<int> v {4,5,6,7,0,1,2};
    int t = 7;
    int res = s.search(v, t);
    std::cout << res << "\n";
    return 0;
}
